File: solutions/1376. Time Needed to Inform All Employees.cpp
class Solution {
 public:
  // Uses dynamic programming to get the informed time of all employees from
  // bottom up.
  int numOfMinutes(int n, int headID, std::vector<int>& manager,
                   std::vector<int>& informTime) {
    head_id_ = headID;
    manager_ = manager;
    inform_time_ = informTime;
    informed_time_ = std::vector<int>(n, NO_CACHE);
    informed_time_.at(headID) = 0;

    int max_time = 0;
    for (int employee = 0; employee < n; employee++) {
      max_time = std::max(GetInformedTime(employee), max_time);
    }
    return max_time;
  }

 private:
  int head_id_;
  std::vector<int> manager_;
  /// @brief The time a manager needs to inform the subordinates.
  std::vector<int> inform_time_;
  int NO_CACHE = -1;
  /// @brief Cached table. The time an employee is informed.
  std::vector<int> informed_time_;

  int GetInformedTime(int employee) {
    if (HasCachedInformedTime_(employee)) {
      return informed_time_.at(employee);
    }
    informed_time_.at(employee) = inform_time_.at(manager_.at(employee))
                                  + GetInformedTime(manager_.at(employee));
    return informed_time_.at(employee);
  }

  bool HasCachedInformedTime_(int employee) const {
    return informed_time_.at(employee) != NO_CACHE;
  }
};

File: solutions/1376. Time Needed to Inform All Employees.rs
impl Solution {
  pub fn num_of_minutes(n: i32, head_id: i32, manager: Vec<i32>, inform_time: Vec<i32>) -> i32 {
      // Cached table. The time an employee is informed.
      let mut informed_time = vec![-1; n as usize];
      informed_time[head_id as usize] = 0;
      let mut max_time: i32 = 0;
      for employee in 0..n {
          max_time = max_time.max(Self::get_informed_time(
              employee,
              &manager,
              &inform_time,
              &mut informed_time,
          ));
      }
      max_time
  }

  fn get_informed_time(
      employee: i32,
      manager: &Vec<i32>,
      inform_time: &Vec<i32>,
      informed_time: &mut Vec<i32>,
  ) -> i32 {
      // has cache
      if informed_time[employee as usize] != -1 {
          return informed_time[employee as usize];
      }
      // compute & add cache
      informed_time[employee as usize] = inform_time[manager[employee as usize] as usize]
          + Self::get_informed_time(
              manager[employee as usize],
              manager,
              inform_time,
              informed_time,
          );
      informed_time[employee as usize]
  }
}
